src/util/timelock.rs

```diff
@@ -0,0 +1,339 @@
+// Rust Bitcoin Library
+// Written in 2022 by
+//     Tobin C. Harding <me@tobin.cc>
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Bitcoin timelocks.
+//!
+//! Relative and absolute timelocks based on `nSequence` and `nLockTime`.
+//!
+
+use core::fmt;
+
+use crate::blockdata::constants::{
+    LOCKTIME_THRESHOLD, SEQUENCE_LOCKTIME_DISABLE_FLAG, SEQUENCE_LOCKTIME_GRANULARITY,
+    SEQUENCE_LOCKTIME_TYPE_FLAG,
+};
+
+/// An absolute timelock, expires after either block height or time.
+///
+/// This is nLockTime as defined in BIP-65. There are two types of nLockTime: lock-by-blockheight
+/// and lock-by-blocktime, distinguished by whether nLockTime < LOCKTIME_THRESHOLD.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Absolute {
+    /// A block height absolute timelock.
+    Height(Height),
+    /// A UNIX timestamp absolute timelock.
+    Time(Time),
+}
+
+impl Absolute {
+    /// Constructs an absolute timestamp from an nLockTime value.
+    pub fn from_n_lock_time(n: u32) -> Self {
+        if encodes_height(n) {
+            Self::Height(Height::new(n).expect("n is valid"))
+        } else {
+            Self::Time(Time::new(n).expect("n is valid"))
+        }
+    }
+
+    /// Returns true if both timelocks are the same unit i.e., both block based or both time based.
+    pub fn is_same_unit(&self, other: &Self) -> bool {
```

Comment:
&self instead of &Self

Reply:
I'm pretty sure this is correct as `Self`, its the type of `self` which special, can only have one of those :)

Reply:
Resolved by using `LockTime` instead of `Self` :)